### PR TITLE
Return result of enrollment test (level) without writing the answers to database

### LINKED ISSUE #89: Get results of enrollment test without writing answers to database
In the current implementation, backend returns level of English as a response to a POST request creating records for `EnrollmentTestAnswer` linked to the given student. However, when the student is taken an assessment, they are still in the middle of the registration process, and their record will not be created for a while after that (they need to answer some additional questions or even take another assessment). It even can't be created at that stage because some required fields of `Student` cannot be filled yet.

That means we need a view that just gets IDs of `EnrollmentTestAnswer` and returns level, nothing more.

The existing view will still be used to create records for the answers later on (maybe without returning the level as this won't be needed anymore).

### PR DESCRIPTION
**Pull request by lemontree210**
closes #89

### CODE REVIEW

File: api/views/enrollment_test.py [hunk 1]
```diff
@@ -16,4 +24,20 @@
...
+    def get_level(self, request: Request) -> Response:
+        """Calculates level of language based on number of correct answers."""
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        answer_ids = cast(list[int], request.POST.getlist("answers"))
+        try:
+            return Response(EnrollmentTestResultLevelSerializer.calculate_level(answer_ids))
+        except NotImplementedError as e:
+            raise ValidationError(str(e))
```
[the remark this answers is not in the document]
Comment (lemontree210) [reply api/views/enrollment_test.py]: Fixed

File: api/models/enrollment_tests.py [hunk 2]
```diff
@@ -81,11 +83,34 @@ def __str__(self) -> str:
     @property
     def resulting_level(self) -> str:
         """Returns language level of the student based on amount of correct answers."""
-        correct_answers = self.answers.filter(is_correct=True).count()
-        total_answers = self.answers.count()
-        percentage = correct_answers / total_answers * 100
-        closest_percentage = min(
-            LEVEL_BY_PERCENTAGE_OF_CORRECT_ANSWERS, key=lambda x: abs(x - percentage)
-        )
+        return self.calculate_level([answer.id for answer in self.answers.all()])
```
Comment (Roxe322) [inline api/models/enrollment_tests.py:86]: self.answers.values_list('id', flat=True)

File: api/models/enrollment_tests.py [hunk 3]
```diff
@@ -81,11 +83,34 @@ def __str__(self) -> str:
...
+
+    @staticmethod
+    def calculate_level(answer_ids: Sequence[int]) -> str:
+        """Calculates language level depending on amount of correct answers.
+
+        This method can be called from outside without creating any `EnrollmentTestResult` objects.
+        """
+        # depending on number of questions in test, the thresholds are different
+        thresholds_for_number_of_questions = {
```
Comment (Roxe322) [inline api/models/enrollment_tests.py:95]: Maybe let's save it as a constant outside the function?

File: api/views/enrollment_test.py [hunk 4]
```diff
@@ -16,4 +24,24 @@ class EnrollmentTestViewSet(ReadOnlyModelViewSet[EnrollmentTest]):
...
+            return EnrollmentTestResultLevelSerializer
+        return EnrollmentTestResultCreateSerializer
+
+    @action(detail=False, methods=["post"])
+    def get_level(self, request: Request) -> Response:
+        """Calculates level of language based on number of correct answers."""
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        answer_ids = cast(list[int], request.POST.getlist("answers"))
```
Comment (Roxe322) [inline api/views/enrollment_test.py:39]: Why don't you use serializer.validated_data here?

File: api/views/enrollment_test.py [hunk 5]
```diff
@@ -16,4 +24,24 @@ class EnrollmentTestViewSet(ReadOnlyModelViewSet[EnrollmentTest]):
...
+        try:
+            return Response(EnrollmentTestResultLevelSerializer.calculate_level(answer_ids))
+        except NotImplementedError:
+            # Repeating the message rather than putting str(e), otherwise CodeQL complains about
```
Comment (Roxe322) [inline api/views/enrollment_test.py:43]: I think we could remove this comment, it's pretty clear why we do any rewording here

File: api/models/enrollment_tests.py [hunk 6]
```diff
@@ -81,11 +83,34 @@ def __str__(self) -> str:
...
-        closest_percentage = min(
-            LEVEL_BY_PERCENTAGE_OF_CORRECT_ANSWERS, key=lambda x: abs(x - percentage)
-        )
+        return self.calculate_level([answer.id for answer in self.answers.all()])
+
+    @staticmethod
+    def calculate_level(answer_ids: Sequence[int]) -> str:
+        """Calculates language level depending on amount of correct answers.
+
+        This method can be called from outside without creating any `EnrollmentTestResult` objects.
```
Comment (Roxe322) [inline api/models/enrollment_tests.py:92]: It's clear because it's staticmethod :)

File: api/models/enrollment_tests.py [hunk 7]
```diff
@@ -81,11 +83,34 @@ def __str__(self) -> str:
...
+        """
+        # depending on number of questions in test, the thresholds are different
+        thresholds_for_number_of_questions = {
+            25: {5: "A1", 11: "A2", 19: "B1"},
```
Comment (Roxe322) [inline api/models/enrollment_tests.py:96]: Also I really think we should make an enum with levels ids

File: api/serializers/enrollment_test.py [hunk 8]
```diff
@@ -50,3 +52,16 @@ class Meta:
 
     # TODO Think about some validation, e.g. answers should be unique for each question
     #  and should be in options for this question
+
+
+class EnrollmentTestResultLevelSerializer(serializers.ModelSerializer[EnrollmentTestResult]):
+    """A serializer used to get level of language based on how many answers are correct."""
+
+    @staticmethod
+    def calculate_level(answer_ids: Sequence[int]) -> dict[str, str]:
```
Comment (Roxe322) [inline api/serializers/enrollment_test.py:61]: I think it should be SerializerMethodField. This field will be read-only, and `answers` should be write-only, so in view you will be able to general staff like `validate` and then `return serializer.data`

**Review by Roxe322 — CHANGES_REQUESTED** [review]

File: api/models/enrollment_tests.py (hunk 2 above)
Comment (lemontree210) [reply api/models/enrollment_tests.py:86]: Done. Yes, I'm not that familiar with QuerySet API yet. And damn how docstrings are sparse in Django!

File: api/models/enrollment_tests.py (hunk 3 above)
Comment (lemontree210) [reply api/models/enrollment_tests.py:95]: Done

File: api/views/enrollment_test.py (hunk 5 above)
Comment (lemontree210) [reply api/views/enrollment_test.py:43]: Done

File: api/models/enrollment_tests.py (hunk 6 above)
Comment (lemontree210) [reply api/models/enrollment_tests.py:92]: Removed

File: api/models/enrollment_tests.py (hunk 7 above)
Comment (lemontree210) [reply api/models/enrollment_tests.py:96]: Done

File: api/views/enrollment_test.py (hunk 4 above)
Comment (lemontree210) [reply api/views/enrollment_test.py:39]: Because I forgot this thing existed :)
Done

File: api/serializers/enrollment_test.py (hunk 8 above)
Comment (lemontree210) [reply api/serializers/enrollment_test.py]: Fixed, I think. Wow that took some effort :)

File: api/migrations/0002_data_migration.py [hunk 9]
```diff
@@ -89,8 +89,16 @@ def _write_languages_and_levels(self):
         )
         Language.objects.bulk_create(languages)
 
-        # no C2 at this school
-        levels = (Level(id=id_) for id_ in ("A0", "A1", "A2", "B1", "B2", "C1"))
+        levels = (
+            Level(id=id_) for id_ in (
```
Comment (Roxe322) [inline api/migrations/0002_data_migration.py:93]: You are able to iterate just through enum class LanguageLevelId.

`Level(id=id_) for id_ in LanguageLevelId`

File: api/models/enrollment_tests.py [hunk 10]
```diff
@@ -76,16 +76,8 @@ class EnrollmentTestResult(models.Model):
...
-    def resulting_level(self) -> str:
-        """Returns language level of the student based on amount of correct answers."""
-        correct_answers = self.answers.filter(is_correct=True).count()
-        total_answers = self.answers.count()
-        percentage = correct_answers / total_answers * 100
-        closest_percentage = min(
-            LEVEL_BY_PERCENTAGE_OF_CORRECT_ANSWERS, key=lambda x: abs(x - percentage)
+        return (
+            f"Answers to enrollment test by {self.student} "
+            f"({self.answers.filter(is_correct=True).count()} correct "
```
Comment (Roxe322) [inline api/models/enrollment_tests.py:81]: In my opinion, put any logic inside f-string is bad practice.
I could agree with `self.answers.count()`, but this one looks too complex to be inside f-string

File: api/serializers/enrollment_test.py [hunk 11]
```diff
@@ -45,8 +52,43 @@ class Meta:
...
+
+
+class EnrollmentTestResultLevelSerializer(serializers.ModelSerializer[EnrollmentTestResult]):
+    """A serializer used to get level of language based on how many answers are correct."""
+
+    resulting_level = serializers.SerializerMethodField()
+
+    class Meta:
+        model = EnrollmentTestResult
+        fields = ["answers", "resulting_level"]
```
Comment (Roxe322) [inline api/serializers/enrollment_test.py:68]: Just a small optimisation: if something is not subject to change, it's better to use immutable data structure (tuple in this case)

File: api/serializers/enrollment_test.py [hunk 12]
```diff
@@ -45,8 +52,43 @@ class Meta:
...
+
+    def validate(self, attrs: dict[str, Any]) -> dict[str, Any]:
+        if len(attrs["answers"]) not in ENROLLMENT_TEST_LEVEL_THRESHOLDS_FOR_NUMBER_OF_QUESTIONS:
+            raise serializers.ValidationError(
+                f"Enrollment test with {len(attrs['answers'])} questions is not supported"
```
Comment (Roxe322) [inline api/serializers/enrollment_test.py:73]: `len(attrs["answers"])` used twice here, so it's better to put it into variable

File: api/views/enrollment_test.py [hunk 13]
```diff
@@ -16,4 +22,21 @@ class EnrollmentTestViewSet(ReadOnlyModelViewSet[EnrollmentTest]):
...
+            return EnrollmentTestResultLevelSerializer
+        return EnrollmentTestResultCreateSerializer
+
+    @action(detail=False, methods=["post"])
+    def get_level(self, request: Request) -> Response:
+        """Calculates level of language based on number of correct answers."""
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        try:
+            return Response({"resulting_level": serializer.data["resulting_level"]})
```
Comment (Roxe322) [inline api/views/enrollment_test.py:37]: Why not just

`return Response(serializer.data)`

?

Comment (Roxe322) [reply api/views/enrollment_test.py:37]: If you will make field `answers` write-only, it won't appear in response

File: api/views/enrollment_test.py [hunk 14]
```diff
@@ -16,4 +22,21 @@ class EnrollmentTestViewSet(ReadOnlyModelViewSet[EnrollmentTest]):
...
+        except NotImplementedError:
+            raise ValidationError(
+                f"Enrollment test with {len(serializer.validated_data['answers'])} "
```
Comment (Roxe322) [inline api/views/enrollment_test.py:40]: I think here you could use serializer.data to make it shorter

Comment (Roxe322) [reply api/views/enrollment_test.py:40]: Also, this excepting here doesn't look good, I think it should be excepted in serializer's validate method

File: api/migrations/0002_data_migration.py (hunk 9 above)
Comment (lemontree210) [reply api/migrations/0002_data_migration.py:93]: Done

File: api/models/enrollment_tests.py (hunk 10 above)
Comment (lemontree210) [reply api/models/enrollment_tests.py:81]: Done

File: api/serializers/enrollment_test.py (hunk 11 above)
Comment (lemontree210) [reply api/serializers/enrollment_test.py:68]: Done (with mine and yours)

File: api/serializers/enrollment_test.py (hunk 12 above)
Comment (lemontree210) [reply api/serializers/enrollment_test.py:73]: Fixed

File: api/views/enrollment_test.py (hunk 13 above)
Comment (lemontree210) [reply api/views/enrollment_test.py:37]: Cool! Fixed

File: api/views/enrollment_test.py (hunk 14 above)
Comment (lemontree210) [reply api/views/enrollment_test.py:40]: Oh, I actually should have deleted the whole `try...except` bit and just leave `return` statement. The block was there before I implemented `.validate` in the serializer. Removed.

File: api/serializers/enrollment_test.py [hunk 15]
```diff
@@ -45,8 +52,47 @@ class Meta:
...
+class EnrollmentTestResultLevelSerializer(serializers.ModelSerializer[EnrollmentTestResult]):
+    """A serializer used to get level of language based on how many answers are correct."""
+
+    resulting_level = serializers.SerializerMethodField()
+
+    class Meta:
+        model = EnrollmentTestResult
+        fields = ("answers", "resulting_level")
+        extra_kwargs = {
+            "answers": {"write_only": True}  # to exclude from serializer.data in ViewSet
```
Comment (Roxe322) [inline api/serializers/enrollment_test.py:70]: Let's remove this comment :)

Comment (lemontree210) [reply api/serializers/enrollment_test.py]: Done

**Review by Roxe322 — APPROVED** [review]

**Pull request merged**